_posts/2021-04-05-fmf-unified-testing.md

```diff
@@ -0,0 +1,156 @@
+
+I can't remember a single bodhi update in the last year when these tests went green, there was always a bunch of failures; so it really got high time to fix this once and for all.
+
+## Fix: Run distribution tests upstream
+
+The conceptual solution is quite obvious:
+ 1. Pin down the environment where these tests run, and provide a tool to create and use them.
+ 2. Make it trivial to locally run and debug a package's gating tests.
+ 3. Run these gating tests for *every* upstream change (i.e. pull request), in *exactly the same* environment, with *exactly the same* test metadata and configuration.
+
+I'm really happy that all of these finally exist!
+
+## Flexible Medadata Format
+
+[FMF](https://tmt.readthedocs.io/en/latest/spec.html) is the successor of the Ansible based [Standard Test Interface](https://docs.fedoraproject.org/en-US/ci/standard-test-interface/). It is declarative, YAML, completely distribution/project agnostic, and rather rich (i.e. flexible), so that it does not limit by design what your tests can do or where to run them. Despite its complexity, most of its settings have sane defaults, so that you don't need to know that much about it.
+
+As a reference for the Cockpit world, check out how FMF was [added to starter-kit](https://github.com/cockpit-project/starter-kit/commit/09823650e222da0): The central file is [test/browser/main.fmf](https://github.com/cockpit-project/starter-kit/blob/master/test/browser/main.fmf) which lists the test dependencies, the entry script, and a timeout:
+
+```yaml
+summary:
+    Run browser integration tests on the host
+require:
+  - cockpit-starter-kit
+  [...]
+  - npm
+  - python3
+test: ./browser.sh
+duration: 60m
+```
+
+This is a pretty straightforward translation from the [STI Ansible tests.yml](https://src.fedoraproject.org/rpms/cockpit/blob/d6853f04e1184f562f872c5c765f40644ef5edc8/f/tests/tests.yml) which looked like this:
+
+```yaml
+- hosts: localhost
+  roles:
+  - role: standard-test-source
+    tags:
+    - always
+
+  - role: standard-test-basic
+    tags:
+    - classic
+    required_packages:
+    - cockpit
+    [...]
+    - npm
+    - python3
+    tests:
+    - verify:
+        dir: .
+        run: ./verify.sh
+        save-files: ["logs/*"]
+```
+
+Aside from that you just need some boilerplate `.fmf/version` (just "1") and at least one top-level [plans/*.fmf](https://github.com/cockpit-project/starter-kit/blob/master/plans/all.fmf) (which can really be the same for every project, and hopefully it will be the implied default some day).
```

Comment:
@TomasTomecek What do you think about making these the default?

Reply:
That's rather a question for @psss I think.

Reply:
Agreed, it would indeed be nice if fmf didn't require this to be set - especially since the version didn't change yet :)